### DISCUSSION TITLE
Can I develop a module for switching outfits among VRM models?

### DISCUSSION THREAD

**Discussion by tejasnayak25** [pixiv/three-vrm#1220] [General]
Hello, 
I've been planning on developing a javascript module for equipping a VRM character with another character's outfit, basically changing outfits.
It only works on similar sized models though.
I'm almost done with the module idea.
I just wanted to know if Pixiv allows me to build something like that. Can I continue? Can I make it into a npm package?
Does Pixiv apply any restrictions on such stuff?
Please let me know.
Thank you.

**Comment by ke456-png** [comment]
I think it's a great idea.

 However, if you intend to use models exported from VRoid Studio, we ask that you follow the guidelines below.

https://vroid.com/en/studio/guidelines#:~:text=When%20creating%20character,and/or%20textures.
> You cannot use VRoid Studio to create an application that can deform meshes and/or can create 3D models like avatars and items by combining meshes and/or textures.

**Reply by tejasnayak25** [reply]
@ke456-png , Thanks for your kind information.

1) I'm not creating a character creation software, I'm creating a javascript module to switch outfits among VRM models.
2) It isn't deforming the mesh, it just applies the outfit mesh of one model on another model.
3) I'm creating this module so that it can help game and software developers to switch outfits among VRM models instead of having to create a separate VRM outfit model for each model. This will cause lots of memory usage. Imagine you have 2 characters and you want them to have 2 similar outfits. Then the user will have to create 4 models, each character wearing two different outfits.
4) This module isn't for creating avatars or 3d models. It just applies the new outfit on the preloaded VRM file in user's project with respect to the provided path for the new outfit VRM.
5) The way in which this module may be used wholly depends on the user.

Am I offending any of the guidelines? If yes, I'll cancel my plans to develop the same.

Kindly let me know. 

Thank you

**Reply by ke456-png** [reply]
VRoid Studio is just one of the applications for creating VRM.
I think it is acceptable to develop a costume editing application that supports the VRM file format.

Since this is the repository of `three-vrm`, I cannot provide any further answers about the specifications or guidelines of VRoid Studio.

**Reply by tejasnayak25** [reply]
@ke456-png , thank you. I would like to know what other applications export VRM models. Do they have the same structure (bones, materials, etc)? Do they allow such programs?

**Reply by ke456-png** [reply]
For information on VRM specifications and applications, please refer to the following links👍 

[https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0/README.md](https://github.com/vrm-c/vrm-specification/blob/master/specification/VRMC_vrm-1.0/README.md)

[https://vrm.dev/en/vrm/vrm_applications.html](https://vrm.dev/en/vrm/vrm_applications.html)
